lib/std/special/compiler_rt/emutls.zig

```diff
@@ -0,0 +1,356 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2015-2018 LLVM Compiler Infrastructure
+// Copyright (c) 2020 Sebastien Marie <semarie@online.fr>
+// This file is part of [zig](https://ziglang.org/), which is MIT licensed.
+// The MIT license requires this copyright notice to be included in all copies
+// and substantial portions of the software.
+// __emutls_get_address specific builtin
+//
+// derived work from LLVM Compiler Infrastructure - release 8.0 (MIT)
+// https://github.com/llvm-mirror/compiler-rt/blob/release_80/lib/builtins/emutls.c
+//
+
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+
+comptime {
+    assert(std.builtin.link_libc);
+    assert(std.Thread.use_pthreads);
+}
+
+/// public entrypoint for generated code using EmulatedTLS
+pub fn __emutls_get_address(control: *emutls_control) callconv(.C) *c_void {
+    const allocator = std.heap.c_allocator; // XXX
+
+    return control.get_pointer(allocator) catch {
+        @panic("emutls_control.get_pointer");
+    };
+}
+
+/// Simple array of ?ObjectPointer with automatic resizing and
+/// automatic storage allocation.
+const ObjectArray = struct {
+    const ObjectPointer = *c_void;
+
+    // content of the array
+    slots: []?ObjectPointer,
+    allocator: *Allocator,
+
+    /// create a new ObjectArray with n slots. must call deinit() to deallocate.
+    pub fn initAlloc(allocator: *Allocator, n: usize) !*ObjectArray {
+        var array = try allocator.create(ObjectArray);
+        errdefer allocator.destroy(array);
+
+        array.* = ObjectArray{
+            .slots = try allocator.alloc(?ObjectPointer, n),
+            .allocator = allocator,
+        };
+        errdefer allocator.free(array.slots);
+
+        for (array.slots) |_object, i| {
+            array.slots[i] = null;
+        }
+
+        return array;
+    }
+
+    /// deallocate the ObjectArray.
+    pub fn deinit(self: *ObjectArray) void {
+        const allocator = self.allocator;
+
+        // deallocated used objects in the array
+        for (self.slots) |_object, i| {
+            if (self.slots[i] != null) {
+                // XXX slot allocated using posix_memalign
+                std.c.free(self.slots[i]);
+                self.slots[i] = null;
+            }
+        }
+        allocator.free(self.slots);
+        self.* = undefined;
+        allocator.destroy(self);
+    }
+
+    /// resize the ObjectArray if needed.
+    pub fn ensureSize(self: *ObjectArray, new_size: usize) !void {
+        const old_size = self.slots.len;
+
+        if (old_size > new_size)
+            return;
+
+        // reallocate
+        self.slots = try self.allocator.reallocAtLeast(self.slots, new_size);
+
+        // init newly added slots
+        for (self.slots[old_size..]) |_object, i| {
+            self.slots[old_size + i] = null;
```

Comment:
```suggestion
for (self.slots[old_size..]) |*object| {
            object.* = null;
}
```

Ditto for the other loops.